warehouse/query-core/src/main/java/datawave/query/jexl/visitors/QueryPruningVisitor.java

```diff
@@ -0,0 +1,500 @@
+        for (int i = 0; i < node.jjtGetNumChildren(); i++) {
+            states.add((TRUTH_STATE) (node.jjtGetChild(i).jjtAccept(this, data)));
+        }
+        
+        if (states.contains(TRUTH_STATE.UNKNOWN) || states.size() > 1) {
+            return TRUTH_STATE.UNKNOWN;
+        } else {
+            if (states.iterator().next() == TRUTH_STATE.TRUE) {
+                replaceAndAssign(node, new ASTFalseNode(ParserTreeConstants.JJTFALSENODE));
+                return TRUTH_STATE.FALSE;
+            } else {
+                replaceAndAssign(node, new ASTTrueNode(ParserTreeConstants.JJTTRUENODE));
+                return TRUTH_STATE.TRUE;
+            }
+        }
+    }
+    
+    @Override
+    public Object visit(ASTOrNode node, Object data) {
+        Set<TRUTH_STATE> states = new HashSet<>();
+        for (int i = 0; i < node.jjtGetNumChildren(); i++) {
+            states.add((TRUTH_STATE) (node.jjtGetChild(i).jjtAccept(this, data)));
+        }
+        
+        if (states.contains(TRUTH_STATE.TRUE)) {
+            replaceAndAssign(node, new ASTTrueNode(ParserTreeConstants.JJTTRUENODE));
+            return TRUTH_STATE.TRUE;
+        } else if (states.contains(TRUTH_STATE.UNKNOWN)) {
+            return TRUTH_STATE.UNKNOWN;
+        } else {
+            replaceAndAssign(node, new ASTFalseNode(ParserTreeConstants.JJTFALSENODE));
+            return TRUTH_STATE.FALSE;
+        }
+    }
+    
+    @Override
+    public Object visit(ASTAndNode node, Object data) {
+        Set<TRUTH_STATE> states = new HashSet<>();
+        for (int i = 0; i < node.jjtGetNumChildren(); i++) {
+            states.add((TRUTH_STATE) (node.jjtGetChild(i).jjtAccept(this, data)));
+        }
+        
+        if (states.contains(TRUTH_STATE.FALSE)) {
+            replaceAndAssign(node, new ASTFalseNode(ParserTreeConstants.JJTFALSENODE));
+            return TRUTH_STATE.FALSE;
+        } else if (states.contains(TRUTH_STATE.UNKNOWN)) {
+            return TRUTH_STATE.UNKNOWN;
+        } else {
+            replaceAndAssign(node, new ASTTrueNode(ParserTreeConstants.JJTTRUENODE));
+            return TRUTH_STATE.TRUE;
+        }
+    }
+    
+    // deal with wrappers
+    @Override
+    public Object visit(ASTJexlScript node, Object data) {
+        // attempt to reduce
+        Set<TRUTH_STATE> states = new HashSet<>();
+        for (int i = 0; i < node.jjtGetNumChildren(); i++) {
+            states.add((TRUTH_STATE) (node.jjtGetChild(i).jjtAccept(this, data)));
+        }
+        
+        if (states.size() == 1) {
+            return states.iterator().next();
+        } else {
+            return TRUTH_STATE.UNKNOWN;
+        }
+    }
+    
+    @Override
+    public Object visit(ASTReference node, Object data) {
+        Set<TRUTH_STATE> states = new HashSet<>();
+        for (int i = 0; i < node.jjtGetNumChildren(); i++) {
+            states.add((TRUTH_STATE) (node.jjtGetChild(i).jjtAccept(this, data)));
```

Comment:
should only be one child.  Lets put an assert everywhere to ensure we only have one child where appropriate.  Then we will not need to create a new HashSet except for OR and AND nodes.

Reply:
agree